File: lib/padrones.js
/* Averiguar el padron desde la direccion. SOLO MONTEVIDEO.

   La Intendencia publica "Direcciones oficiales de Montevideo": 377.583 puntos con calle,
   numero de puerta y padron. No hay ningun servicio al que preguntarle de a uno, asi que
   el robot se baja la tabla entera y arma el indice (robot/padrones.py). Aca solo se
   consulta.

   Los otros departamentos no publican nada equivalente. Cuando la direccion es de otro
   lado esto lo dice y ofrece el visor oficial — no adivina. Un padron equivocado
   identifica OTRA propiedad en un documento que obliga.

   Da el padron del EDIFICIO, no del apartamento: la tabla no tiene esa columna. En
   propiedad horizontal la carta se escribe "padron NNNN, unidad 202", asi que alcanza. */

const GRUPOS = 48;
export const DEPARTAMENTO_CUBIERTO = "Montevideo";
/* En http a proposito: el visor NO responde por https. Ver la nota de papelesDeCatastro. */
export const VISOR_CATASTRO = "http://visor.catastro.gub.uy/visordnc/";

/* Para comparar: "MAROÑAS" y "MARONAS" tienen que encontrarse igual. */
export const sinTildes = (texto) =>
  String(texto).toUpperCase().normalize("NFD").replace(/[̀-ͯ]/g, "");

export const normalizar = (calle) =>
  sinTildes(calle).replace(/\./g, " ").replace(/\s+/g, " ").trim();

/* La MISMA cuenta que robot/padrones.py. Si las dos no dan el mismo numero, la app busca
   en el archivo equivocado y no encuentra nada — por eso hay un test que la fija. */
export function grupoDe(calle) {
  let h = 0;
  for (const letra of normalizar(calle)) h = (h * 31 + letra.codePointAt(0)) % 1000003;
  return String(h % GRUPOS).padStart(2, "0");
}

/* Los archivos se piden una vez y quedan. Son 29 KB comprimidos cada uno. */
const enMemoria = new Map();

/* Por defecto se pide por la red. `leer` existe para poder probarlo sin servidor: los
   tests le pasan una funcion que abre el archivo del disco. */
async function porLaRed(grupo, base) {
  const respuesta = await fetch(new URL(`datos/padrones/${grupo}.json`, base));
  if (!respuesta.ok) throw new Error(grupo);
  return respuesta.json();
}

async function traerGrupo(calle, base, leer) {
  const grupo = grupoDe(calle);
  if (enMemoria.has(grupo)) return enMemoria.get(grupo);
  try {
    const datos = leer ? await leer(grupo) : await porLaRed(grupo, base);
    if (!datos) return null;
    if (!leer) enMemoria.set(grupo, datos);
    return datos;
  } catch {
    return null;
  }
}

/* "1881:82447;1885:82448" -> Map con los numeros de puerta y su padron. */
function comoMapa(compacto) {
  const puertas = new Map();
  for (const par of String(compacto).split(";")) {
    const corte = par.indexOf(":");
    if (corte > 0) puertas.set(Number(par.slice(0, corte)), par.slice(corte + 1));
  }
  return puertas;
}

/* Los numeros de puerta mas cercanos que SI existen. Sirve para no dejarlo en la nada:
   RE/MAX publica muchas direcciones redondeadas a la cuadra ("Eusebio Vidal 3100") y ese
   numero no existe como puerta. */
function cercanos(puertas, numero, cuantos = 4) {
  return [...puertas.keys()]
    .sort((a, b) => Math.abs(a - numero) - Math.abs(b - numero) || a - b)
    .slice(0, cuantos)
    .sort((a, b) => a - b)
    .map((n) => ({ numero: n, padron: puertas.get(n) }));
}

/* Devuelve siempre un objeto con `estado`, para que la pantalla no tenga que adivinar:
     "fuera-de-montevideo" | "sin-datos" | "calle-desconocida" | "sin-numero" | "encontrado" */
export async function buscar({ calle, numero, departamento }, { base = "", leer = null } = {}) {
  const donde = base || (typeof window !== "undefined" ? window.location.href : "");

  if (departamento && normalizar(departamento) !== normalizar(DEPARTAMENTO_CUBIERTO)) {
    return { estado: "fuera-de-montevideo", departamento };
  }
  if (!String(calle || "").trim()) return { estado: "calle-desconocida", calle };

  const grupo = await traerGrupo(calle, donde, leer);
  if (!grupo) return { estado: "sin-datos" };

  const compacto = grupo[normalizar(calle)];
  if (!compacto) return { estado: "calle-desconocida", calle };

  const puertas = comoMapa(compacto);
  const puerta = Number(String(numero).replace(/\D/g, ""));
  if (!puerta) return { estado: "sin-numero", calle, puertas: puertas.size };

  const padron = puertas.get(puerta);
  if (padron) return { estado: "encontrado", padron, calle, numero: puerta };

  return { estado: "sin-numero-exacto", calle, numero: puerta, cercanos: cercanos(puertas, puerta) };
}

/* El nombre de calle mas parecido a lo que se escribio, para sugerir mientras tipea. */
export function sugerir(calles, escrito, cuantas = 6) {
  const busca = normalizar(escrito);
  if (busca.length < 2) return [];
  const empiezan = [];
  const contienen = [];
  for (const calle of calles) {
    const plana = normalizar(calle);
    if (plana.startsWith(busca)) empiezan.push(calle);
    else if (plana.includes(busca)) contienen.push(calle);
    if (empiezan.length >= cuantas) break;
  }
  return [...empiezan, ...contienen].slice(0, cuantas);
}

/* Los papeles oficiales de Catastro para un padron.

   Salen del visor geoCatastro. No hay documentacion: las direcciones se sacaron leyendo el
   javascript del propio visor (`CedulaCatastral.js`) y estan PROBADAS contra el servidor,
   una por una. Montevideo es el departamento "V" y la localidad catastral "AA"; como esta
   herramienta solo cubre Montevideo, van fijos.

   TRES COSAS QUE COSTARON Y NO HAY QUE VOLVER A PERDER:

   1. Va por **https en el puerto 8443**. El mismo servlet existe en http:8080 y devuelve
      lo mismo, pero la app se sirve por https y el navegador no abre un http en un puerto
      raro sin plantar antes una pantalla de peligro. Para el usuario eso es "no anda".

   2. El **visor no tiene https**: `https://visor.catastro.gub.uy` no contesta nada, ni
      redirige. Va en http o no va.

   3. La **"unidad" de Catastro NO es el numero de apartamento.** Probado contra el padron
      82447: la unidad 1 devuelve la cedula y la 202 devuelve un archivo vacio. Catastro
      numera las unidades a su manera. Por eso la unidad se PREGUNTA y no se adivina —
      inventarla devuelve el papel de otra unidad, sin decir nada.

   El croquis de manzana y el listado de planos **no se pueden armar desde aca**, y esta
   medido, no supuesto: el croquis se pide por SOAP (`wsVisor.asmx`, un solo metodo, sin
   variante GET) y necesita el numero de manzana; la manzana sale de un servicio de mapas
   que existe solo en http, y desde una pagina https el navegador bloquea esa consulta
   siempre. Para esos dos se manda al visor, que los tiene a un clic. */
const SERVLET = "https://apls2.catastro.gub.uy:8443/integralevol3produccion/servlet/"
  + "gub.catastro.integralevol3produccion";

const TERRITORIAL = "https://intgis.montevideo.gub.uy/sit/aplicaciones/montevimap/"
  + "formularios/padrones.php";

/* Catastro separa los pedazos de la direccion CON COMAS. Si el usuario escribe una coma
   en la unidad —"202,X"— la direccion queda con ocho pedazos en vez de siete y Catastro
   devuelve otro documento o un error. Se deja pasar solo letras y numeros. */
const soloLetrasYNumeros = (texto) =>
  String(texto || "").normalize("NFD").replace(/[^0-9A-Za-z]/g, "");

export function papelesDeCatastro(padron, { apartamento = "", bloque = "", unidad = "" } = {}) {
  const bloq = soloLetrasYNumeros(bloque);
  const uni = soloLetrasYNumeros(unidad);
  const esApartamento = Boolean(soloLetrasYNumeros(apartamento) || bloq || uni);
  const papeles = [];

  if (esApartamento) {
    /* Los datos de la parcela salen SIN saber la unidad, y adentro vienen listadas todas
       las unidades con su destino y sus metros. Por eso van primero: son el papel que
       responde cual es la unidad que hace falta para pedir la cedula. */
    papeles.push({
      clave: "parcela", nombre: "Datos completos de la parcela", pdf: true,
      url: `${SERVLET}.arwebphmvdeopublico?V,AA,${padron},${bloq},,${uni},N`,
    });
    if (uni) {
      papeles.push({
        clave: "cedula", nombre: `Cédula catastral (unidad ${uni})`, pdf: true,
        url: `${SERVLET}.apwebimpresioncedulasgeocatastro?H,V,AA,${padron},${bloq},,${uni}`,
      });
    }
  } else {
    papeles.push({
      clave: "cedula", nombre: "Cédula catastral", pdf: true,
      url: `${SERVLET}.apwebimpresioncedulasgeocatastro?C,V,AA,${padron},,,`,
    });
    papeles.push({
      clave: "parcela", nombre: "Datos completos de la parcela", pdf: true,
      url: `${SERVLET}.arwebmvdeocomunpublico?${padron},N`,
    });
  }

  papeles.push({
    clave: "territorial", nombre: "Datos territoriales de la Intendencia", pdf: false,
    url: `${TERRITORIAL}?padron=${encodeURIComponent(padron)}`
      + "&servidor=intgis.montevideo.gub.uy&language=es",
  });
  papeles.push({
    clave: "visor", nombre: "Croquis de manzana y planos (en el visor)", pdf: false,
    url: VISOR_CATASTRO,
  });
  return papeles;
}

/* El papel mas util que se puede armar para ese padron: la cedula si se puede pedir, y
   si no los datos de la parcela, que salen siempre. */
export const infoCatastral = (padron, opciones) => {
  const papeles = papelesDeCatastro(padron, opciones);
  return (papeles.find((p) => p.clave === "cedula") || papeles[0]).url;
};

File: robot/padrones.py
"""Los padrones de Montevideo, para poder averiguarlos desde la direccion.

La Intendencia publica "Direcciones oficiales de Montevideo": 377.583 puntos con calle,
numero de puerta y PADRON. No hay ningun servicio al que preguntarle de a uno — hay que
bajarse la tabla entera y armarse el indice. Eso es lo que hace este modulo.

Verificado contra un caso real: Pantaleon Perez 4782 -> padron 62295, que es el que el
usuario habia escrito a mano en una carta oferta de verdad.

SOLO MONTEVIDEO. Los otros departamentos no publican nada equivalente; Canelones tiene un
visor para consultar de a uno y nada descargable. Cuando la direccion es de otro lado, la
app lo dice y ofrece abrir el visor oficial.

El indice se parte en 48 pedazos parejos: el telefono baja solo el que necesita (unos
80 KB) en vez de un archivo de 3,9 MB. Se reparte con una cuenta y no por letra inicial
porque por letra queda desparejo — muchas calles empiezan con "AV" o "CNO" y esos dos
archivos se llevaban la quinta parte del total.
"""

from __future__ import annotations

import json
import pathlib
import re
import struct
import unicodedata
import urllib.request
import zipfile

RAIZ = pathlib.Path(__file__).resolve().parent.parent
DESTINO = RAIZ / "datos" / "padrones"

# El generador arma el zip del lado del servidor y despues redirige a el. Son dos pedidos:
# el primero lo construye, el segundo lo baja.
GENERAR = ("https://intgis.montevideo.gub.uy/sit/php/common/datos/"
           "generar_zip2.php?nom_tab=v_mdg_accesos&tipo=gis")
BAJAR = "https://intgis.montevideo.gub.uy/sit/tmp/v_mdg_accesos.zip"


def sin_tildes(texto: str) -> str:
    """Para comparar. 'MAROÑAS' y 'MARONAS' tienen que encontrarse igual."""
    plano = unicodedata.normalize("NFD", texto.upper())
    return "".join(c for c in plano if unicodedata.category(c) != "Mn")


def normalizar(calle: str) -> str:
    """La forma con la que se busca: sin tildes, sin puntos y sin espacios de mas."""
    return re.sub(r"\s+", " ", sin_tildes(calle).replace(".", " ")).strip()


GRUPOS = 48

def grupo_de(calle: str) -> str:
    """En que archivo cae. La cuenta esta escrita igual en lib/padrones.js: si las dos no
    dan el mismo numero, la app busca en el archivo equivocado y no encuentra nada."""
    h = 0
    for letra in normalizar(calle):
        h = (h * 31 + ord(letra)) % 1000003
    return f"{h % GRUPOS:02d}"


def leer_dbf(ruta: pathlib.Path) -> list[dict]:
    """Lee un .dbf con la biblioteca estandar. Es un formato viejo y simple: una cabecera
    que dice cuantos registros hay y de que largo, y despues los datos de corrido."""
    with ruta.open("rb") as f:
        cabecera = f.read(32)
        registros, largo_cabecera, largo_registro = struct.unpack("<I H H", cabecera[4:12])
        campos = []
        while True:
            c = f.read(32)
            if c[0:1] in (b"\r", b""):
                break
            campos.append((c[:11].split(b"\x00")[0].decode("latin1"), c[16]))
        f.seek(largo_cabecera)
        crudo = f.read(registros * largo_registro)

    posiciones, pos = {}, 1   # el byte 0 de cada registro es la marca de borrado
    for nombre, largo in campos:
        posiciones[nombre] = (pos, largo)
        pos += largo

    filas = []
    for i in range(registros):
        base = i * largo_registro
        if crudo[base:base + 1] == b"*":      # registro borrado
            continue
        fila = {}
        for nombre, (desde, largo) in posiciones.items():
            fila[nombre] = crudo[base + desde:base + desde + largo].decode("utf-8", "replace").strip()
        filas.append(fila)
    return filas


def bajar(destino: pathlib.Path) -> pathlib.Path:
    urllib.request.urlopen(GENERAR, timeout=300).read()      # lo arma del otro lado
    with urllib.request.urlopen(BAJAR, timeout=600) as r:
        destino.write_bytes(r.read())
    return destino


def construir(filas: list[dict]) -> dict[str, dict[str, dict[str, str]]]:
    """De la tabla cruda al indice: por grupo, por calle, numero -> padron."""
    por_grupo: dict[str, dict[str, dict[str, str]]] = {}
    for fila in filas:
        calle = fila.get("NOM_CALLE", "").strip()
        numero = fila.get("NUM_PUERTA", "").strip()
        padron = fila.get("PADRON", "").strip()
        if not (calle and numero and padron):
            continue
        grupo = por_grupo.setdefault(grupo_de(calle), {})
        grupo.setdefault(normalizar(calle), {})[numero] = padron
    return por_grupo


def escribir(por_grupo, nombres_reales: dict[str, str], destino: pathlib.Path = DESTINO) -> dict:
    destino.mkdir(parents=True, exist_ok=True)
    for viejo in destino.glob("*.json"):
        viejo.unlink()

    calles = []
    for grupo, contenido in sorted(por_grupo.items()):
        compacto = {
            calle: ";".join(f"{n}:{p}" for n, p in sorted(puertas.items(), key=lambda x: int(x[0])))
            for calle, puertas in contenido.items()
        }
        (destino / f"{grupo}.json").write_text(
            json.dumps(compacto, ensure_ascii=False, separators=(",", ":")), encoding="utf-8")
        for calle in contenido:
            calles.append(nombres_reales.get(calle, calle))

    calles.sort()
    (destino / "calles.json").write_text(
        json.dumps(calles, ensure_ascii=False, separators=(",", ":")), encoding="utf-8")

    pesos = {p.name: p.stat().st_size for p in destino.glob("*.json")}
    return {
        "calles": len(calles),
        "direcciones": sum(len(v) for g in por_grupo.values() for v in g.values()),
        "archivos": len(pesos),
        "peso_total_kb": round(sum(pesos.values()) / 1024),
        "archivo_mas_grande_kb": round(max(pesos.values()) / 1024),
    }


def actualizar(zip_local: pathlib.Path | None = None) -> dict:
    """Baja (o usa uno ya bajado), arma el indice y lo escribe en datos/padrones/."""
    import tempfile

    with tempfile.TemporaryDirectory() as tmp:
        carpeta = pathlib.Path(tmp)
        z = zip_local or bajar(carpeta / "accesos.zip")
        with zipfile.ZipFile(z) as zf:
            zf.extractall(carpeta)
        dbf = next(p for p in carpeta.iterdir() if p.suffix.lower() == ".dbf")
        filas = leer_dbf(dbf)

    nombres_reales = {}
    for fila in filas:
        calle = fila.get("NOM_CALLE", "").strip()
        if calle:
            nombres_reales.setdefault(normalizar(calle), calle)

    return escribir(construir(filas), nombres_reales)


if __name__ == "__main__":
    import sys
    local = pathlib.Path(sys.argv[1]) if len(sys.argv) > 1 else None
    resumen = actualizar(local)
    for clave, valor in resumen.items():
        print(f"{clave}: {valor}")
